sympy/stats/stochastic_process_types.py

```diff
@@ -759,3 +766,212 @@ def limiting_distribution(self):
+
+        Parameters
+        ==========
+
+        ntrials: Positive Integer,
+          Represents number of trials.
+        rsuccess: Positive Integer including 0, by default is None
+          Represents total number of success events (rsuccess <= ntrials).
+        evaluate: Boolean, by default is False
+          Computes the result numerically.
+
+        Returns
+        =======
+        Proability of getting `r` success in `n` trials
+            If evaluate is True
+        Distribution of getting `r` success in `n` trials
+            If evaluate is False
+
+        """
+        if evaluate:
+            if not (S(ntrials).is_Integer and S(rsuccess).is_Integer):
+                raise ValueError("ntrials and rsuccess must be postive Integers.")
+
+            if not (rsuccess <= ntrials and rsuccess >= 0):
+                raise ValueError("rsuccess <= ntrials and  ntrials, \
+                                     rsuccess must be positive Integers.")
+
+            return self.p**rsuccess * (1 - self.p)**(ntrials - rsuccess)
+
+        else:
+            if not (S(ntrials).is_Integer and ntrials > 0):
+                raise ValueError("ntrials must be a positive Integer.")
+            r = symbols('r', natural=True)
+            pmf = (self.p**r) * (1 - self.p)**(ntrials - r)
+            return DiscreteDistributionHandmade(pmf, set = FiniteSet(*range(ntrials+1)))
+
+    def _rvindexed_subs(self, expr, condition):
+        """
+        Substitutes the RandomIndexedSymbol with the RandomSymbol with
+        same name, distribution and probability as RandomIndexedSymbol.
+
+        """
+        from sympy.stats import Bernoulli
+        rvs_expr = random_symbols(expr)
+        swapdict_expr = {}
+        for rv in rvs_expr:
+            newrv = Bernoulli(rv.name, p=rv.pspace.process.p)
+            swapdict_expr[rv] = newrv
+        expr = expr.subs(swapdict_expr)
+        rvs_cond = random_symbols(condition)
+        swapdict_cond = {}
+        if condition is not None:
+            for rv in rvs_cond:
+                newrv = Bernoulli(rv.name, p=rv.pspace.process.p)
+                swapdict_cond[rv] = newrv
+            condition = condition.subs(swapdict_cond)
+        return expr, condition
+
+    def expectation(self, expr, condition=None, evaluate=True, **kwargs):
+        """
+        Computes expectation.
+
+        Parameters
+        ==========
+
+        expr: RandomIndexedSymbol, Relational, Logic
+            Condition for which expectation has to be computed. Must
+            contain a RandomIndexedSymbol of the process.
+        condition: Relational, Logic
+            The given conditions under which computations should be done.
+
+        Returns
+        =======
+
+        Expectation of the RandomIndexedSymbol.
+
+        """
+        newexpr, newcondition = self._rvindexed_subs(expr, condition)
+        return E(expr=newexpr, condition=newcondition, evaluate=evaluate, **kwargs)
```

Comment:
The following works with the current tests in `test_stochastic_process.py`
```python
newexpr, newcondition = self._rvindexed_subs(expr, condition)
        from sympy.stats import pspace
        new_pspace = pspace(newexpr)
        if newcondition is not None:
            from sympy.stats import given
            newexpr = given(newexpr, newcondition)
        if newexpr.is_Add:  # We know that E is Linear
            return Add(*[new_pspace.compute_expectation(
                        expr=arg, evaluate=evaluate, **kwargs)
                        for arg in newexpr.args])
        return new_pspace.compute_expectation(
                newexpr, evaluate=evaluate, **kwargs)
```
However, I am not sure if it covers all cases. May be readability is affected by saving the CPU cycles. There is a trade off between code cleanliness/re-usability and efficiency. What do you think, should we do?

Reply:
Yeah, This seems to work good as we have cut off 2-3 function calls from the cycle, and also reused the implemented code of Bernoulli dist. , So I think it is good to implement this way, as it is much similar to the implementation of expectation in rv.py and is easily relatable . The flow of the code is also interpretable that RandomIndexed rv are substituted by Bernoulli rv with same probability and then the same implementation is followed as of for Bernoulli rv.

Reply:
Sure, then I think you can push the changes.